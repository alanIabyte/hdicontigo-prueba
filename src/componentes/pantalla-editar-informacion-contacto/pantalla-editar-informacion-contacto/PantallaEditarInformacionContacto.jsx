import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { v4 } from "uuid";
import { useHistory } from "react-router-dom";
import { useCookies } from "react-cookie";
import {
  ContenedorBoton,
  ContenedorValor,
  MensajeError,
  MensajePequeno,
  Separador,
  SeparadorEspacio,
  SeparadorLinea,
  Titulo,
} from "./PantallaEditarInformacionContacto.styled";
import {
  EnvolvedorPantalla,
  Pantalla,
} from "../../componentes-styled-compartidos/Pantalla.styled";
import Encabezado from "../../encabezado";
import Boton from "../../boton";
import CampoTexto from "../../campo-texto";
import Constantes from "../../../recursos/constantes";

const valores = {
  telefono: "",
  email: "",
  color: "",
  placas: "",
};

const { nombreDeCookie } = Constantes;

const PantallaEditarInformacionContacto = () => {
  const history = useHistory();
  const estadoApp = useSelector((estado) => estado);
  const dispatch = useDispatch();
  const [cookie] = useCookies([nombreDeCookie]);
  let prellenadoTelefono = false;

  if (estadoApp && estadoApp.informacionContacto) {
    valores.telefono = estadoApp.informacionContacto.telefono;
    valores.email = estadoApp.informacionContacto.email;
    valores.color = estadoApp.informacionContacto.color;
    valores.placas = estadoApp.informacionContacto.placas;
    const objetoCookie = cookie[nombreDeCookie];
    if (objetoCookie && valores.telefono === objetoCookie.Telefono) {
      prellenadoTelefono = true;
    }
  }

  const diccionario = {
    encabezado: "Editar",
    titulo: "Por favor verifica y completa la siguiente información",
    mensajePequeno: "Estos datos le facilitarán a tu ajustador localizarte.",
    separadorInfoContacto: "información de contacto",
    separadorInfoVehiculo: "información del vehículo",
    etiquetas: {
      telefono: "Teléfono",
      email: "Email",
      color: "Color",
      placas: "Placas",
      botonGuardar: "Guardar cambios",
    },
    marcadores: {
      telefono: "Ingresa tu número de teléfono.",
      email: "Ingresa tu email.",
      color: "Ingresa el color del vehículo.",
      placas: "Ingresa el número de placas.",
    },
    errores: {
      campoRequerido: "Campo requerido para poder continuar.",
      emailFormato: "El correo electrónico no tiene un formato válido.",
    },
  };

  let esTradicional = false;

  if (
    estadoApp.datosIngresoPoliza &&
    estadoApp.datosIngresoPoliza.fechaPasada
  ) {
    // Si el accidente ocurrió en un día previo, el cuestionario y el
    // estado del semáforo se ignoran, el ajustador debe ser digital y
    // por lo mismo el correo debe ser obligatorio.
    esTradicional = false;
  } else if (
    estadoApp.semaforoAmarillo ||
    (estadoApp.cuestionarioReporte &&
      (estadoApp.cuestionarioReporte.respuesta1 || // "¿Hay personas lesionadas?"no
        estadoApp.cuestionarioReporte.respuesta2 || // "¿Hay otro vehículo involucrado u otros bienes afectados?"no
        estadoApp.cuestionarioReporte.respuesta3 || // "¿Hay autoridades presentes?"no
        estadoApp.cuestionarioReporte.respuesta4 || // "¿El vehículo está derramando líquido?"no
        // estadoApp.cuestionarioReporte.respuesta5 || // "¿El siniestro sucedió en ciudad/zona urbana?"
        estadoApp.cuestionarioReporte.respuesta7 || // "¿En el tablero hay indicadores encendidos?"no
        !estadoApp.cuestionarioReporte.respuesta8 || // "¿La unidad puede circular por sus propios medios?"si
        estadoApp.datosPoliza.contieneModulos))
  ) {
    // Si el incidente fue hoy, y una de las primeras siete respuestas del cuestionario
    // es SÍ o el usuario se saltó el cuestionario por ser semáforo amarillo, se
    // usará el ajustador tradicional y el correo no será obligatorio.
    esTradicional = true;
  }

  const [focoTelefono, asignarFocoTelefono] = useState("");
  const [focoEmail, asignarFocoEmail] = useState("");
  const [focoColor, asignarFocoColor] = useState("");
  const [focoPlacas, asignarFocoPlacas] = useState("");
  const [errorTelefono, asignarErrorTelefono] = useState("");
  const [errorEmail, asignarErrorEmail] = useState("");
  const [errorColor, asignarErrorColor] = useState("");
  const [errorPlacas, asignarErrorPlacas] = useState("");

  const alCambiarTelefono = (evento) => {
    if (evento) {
      const valor = evento.target.value;
      valores.telefono = valor;
    }
  };

  const alCambiarEmail = (evento) => {
    if (evento) {
      const valor = evento.target.value;
      valores.email = valor;
    }
  };

  const alCambiarColor = (evento) => {
    if (evento) {
      const valor = evento.target.value;
      valores.color = valor;
    }
  };

  const alCambiarPlacas = (evento) => {
    if (evento) {
      const valor = evento.target.value;
      valores.placas = valor;
    }
  };

  const validarFormulario = () => {
    const { telefono, email, color, placas } = valores;
    let validado = true;

    if (!telefono) {
      validado = false;
      asignarFocoTelefono("error");
      asignarErrorTelefono(diccionario.errores.campoRequerido);
    } else {
      asignarFocoTelefono("");
      asignarErrorTelefono("");
    }

    if (!email && !esTradicional) {
      validado = false;
      asignarFocoEmail("error");
      asignarErrorEmail(diccionario.errores.campoRequerido);
    } else if (
      email &&
      !/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,4})+$/.test(email)
    ) {
      validado = false;
      asignarFocoEmail("error");
      asignarErrorEmail(diccionario.errores.emailFormato);
    } else {
      asignarFocoEmail("");
      asignarErrorEmail("");
    }

    if (!placas && !esTradicional) {
      asignarFocoPlacas("error");
      asignarErrorPlacas(diccionario.errores.campoRequerido);
    } else {
      asignarFocoPlacas("");
      asignarErrorPlacas("");
    }

    if (!color) {
      validado = false;
      asignarFocoColor("error");
      asignarErrorColor(diccionario.errores.campoRequerido);
    } else {
      asignarFocoColor("");
      asignarErrorColor("");
    }

    if (validado === true) {
      history.push("/resumen-reporte");
      dispatch({
        type: "AGREGAR",
        valor: { telefono, email, color, placas },
        indice: "informacionContacto",
      });
    }
  };

  return (
    <EnvolvedorPantalla key={v4()} id="EditarInformacionContacto">
      <Encabezado
        titulo={diccionario.encabezado}
        funcionRegresar={() => {
          history.push("/resumen-reporte");
        }}
        alertaAmarilla={estadoApp && estadoApp.semaforoAmarillo}
      />
      <Pantalla>
        <Titulo id="titulo">{diccionario.titulo}</Titulo>
        <MensajePequeno id="mensajePequeno">
          {diccionario.mensajePequeno}
        </MensajePequeno>
        <Separador id="separadorInfoContacto">
          {diccionario.separadorInfoContacto}
        </Separador>
        <SeparadorLinea />
        {prellenadoTelefono ? (
          <ContenedorValor>{valores.telefono}</ContenedorValor>
        ) : (
          <CampoTexto
            id="campoTelefono"
            etiqueta={diccionario.etiquetas.telefono}
            valor={valores.telefono}
            foco={focoTelefono}
            numeroDeCaracteres={10}
            esNumerico
            enCambio={alCambiarTelefono}
            marcador={diccionario.marcadores.telefono}
          />
        )}
        {errorTelefono !== "" && (
          <MensajeError id="errorTelefono">{errorTelefono}</MensajeError>
        )}
        <SeparadorEspacio />
        <CampoTexto
          id="campoEmail"
          etiqueta={diccionario.etiquetas.email}
          marcador={diccionario.marcadores.email}
          valor={valores.email}
          foco={focoEmail}
          enCambio={alCambiarEmail}
        />
        {errorEmail !== "" && (
          <MensajeError id="errorEmail">{errorEmail}</MensajeError>
        )}
        <SeparadorEspacio />
        <SeparadorEspacio />
        <Separador id="separadorInfoVehiculo">
          {diccionario.separadorInfoVehiculo}
        </Separador>
        <SeparadorLinea />
        <CampoTexto
          id="campoColor"
          etiqueta={diccionario.etiquetas.color}
          marcador={diccionario.marcadores.color}
          valor={valores.color}
          foco={focoColor}
          enCambio={alCambiarColor}
        />
        {errorColor !== "" && (
          <MensajeError id="errorColor">{errorColor}</MensajeError>
        )}
        <SeparadorEspacio />
        <CampoTexto
          id="campoPlacas"
          etiqueta={diccionario.etiquetas.placas}
          marcador={diccionario.marcadores.placas}
          valor={valores.placas}
          numeroDeCaracteres={10}
          foco={focoPlacas}
          enCambio={alCambiarPlacas}
        />
        {errorPlacas !== "" && (
          <MensajeError id="errorPlacas">{errorPlacas}</MensajeError>
        )}
        <ContenedorBoton>
          <Boton
            etiqueta={diccionario.etiquetas.botonGuardar}
            tema="estandar"
            enClick={validarFormulario}
          />
        </ContenedorBoton>
      </Pantalla>
    </EnvolvedorPantalla>
  );
};

export default PantallaEditarInformacionContacto;
