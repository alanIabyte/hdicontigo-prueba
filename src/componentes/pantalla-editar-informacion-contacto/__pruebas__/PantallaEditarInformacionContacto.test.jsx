/* eslint-disable max-len */
import React from "react";
import renderer from "react-test-renderer";
import PantallaEditarInformacionContacto from "../pantalla-editar-informacion-contacto/PantallaEditarInformacionContacto";

jest.mock("react-router-dom", () => ({
  useLocation: jest.fn().mockReturnValue({}),
  useHistory: jest.fn().mockReturnValue({}),
}));

jest.mock("react-redux", () => ({
  useSelector: jest.fn().mockReturnValue({}),
  useDispatch: jest.fn().mockReturnValue({}),
}));

describe("PantallaEditarInformacionContacto", () => {
  it("se despliega correctamente", () => {
    const tree = renderer
      .create(<PantallaEditarInformacionContacto />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
