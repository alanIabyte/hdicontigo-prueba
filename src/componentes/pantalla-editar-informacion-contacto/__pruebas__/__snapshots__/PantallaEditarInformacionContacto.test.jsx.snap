// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PantallaEditarInformacionContacto se despliega correctamente 1`] = `
<div
  className="sc-jsJBEP hrEzkG"
  id="EditarInformacionContacto"
>
  <div>
    Soy un contenedor simulado
  </div>
  <div
    className="sc-koXPp gBlQG"
  >
    <h3
      className="sc-eqUAAy sc-hmdomO hbEPRd gqXMcN"
      id="titulo"
    >
      Por favor verifica y completa la siguiente información
    </h3>
    <div
      className="sc-iGgWBj sc-feUZmu fgQoGq bIzYaA"
      id="mensajePequeno"
    >
      Estos datos le facilitarán a tu ajustador localizarte.
    </div>
    <div
      className="sc-jlZhew sc-fUnMCh iZPiM dEGZCb"
      id="separadorInfoContacto"
    >
      información de contacto
    </div>
    <div
      className="sc-fHjqPf chBSVp"
    />
    <div
      className="sc-cWSHoV iHJLlN"
      id="campoTelefono"
    >
      <div
        className="sc-kpDqfm sc-eBMEME bTCkFB hwrltx"
        id="campoTelefono-etiqueta"
      >
        Teléfono
      </div>
      <div
        className="sc-kpDqfm sc-bmzYkS bTCkFB hLmFXM"
        id="campoTelefono-contenedorCampoDeTexto"
      >
        <input
          autoFocus={false}
          className="sc-iHGNWf jOstXH"
          data-indice={null}
          disabled={false}
          id="campoTelefono-campoDeTexto"
          maxLength={10}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          placeholder="Ingresa tu número de teléfono."
          readOnly={false}
          rows={null}
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      className="sc-hzhJZQ gQJJDh"
    />
    <div
      className="sc-cWSHoV iHJLlN"
      id="campoEmail"
    >
      <div
        className="sc-kpDqfm sc-eBMEME bTCkFB hwrltx"
        id="campoEmail-etiqueta"
      >
        Email
      </div>
      <div
        className="sc-kpDqfm sc-bmzYkS bTCkFB hLmFXM"
        id="campoEmail-contenedorCampoDeTexto"
      >
        <input
          autoFocus={false}
          className="sc-iHGNWf jOstXH"
          data-indice={null}
          disabled={false}
          id="campoEmail-campoDeTexto"
          maxLength=""
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          placeholder="Ingresa tu email."
          readOnly={false}
          rows={null}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="sc-hzhJZQ gQJJDh"
    />
    <div
      className="sc-hzhJZQ gQJJDh"
    />
    <div
      className="sc-jlZhew sc-fUnMCh iZPiM dEGZCb"
      id="separadorInfoVehiculo"
    >
      información del vehículo
    </div>
    <div
      className="sc-fHjqPf chBSVp"
    />
    <div
      className="sc-cWSHoV iHJLlN"
      id="campoColor"
    >
      <div
        className="sc-kpDqfm sc-eBMEME bTCkFB hwrltx"
        id="campoColor-etiqueta"
      >
        Color
      </div>
      <div
        className="sc-kpDqfm sc-bmzYkS bTCkFB hLmFXM"
        id="campoColor-contenedorCampoDeTexto"
      >
        <input
          autoFocus={false}
          className="sc-iHGNWf jOstXH"
          data-indice={null}
          disabled={false}
          id="campoColor-campoDeTexto"
          maxLength=""
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          placeholder="Ingresa el color del vehículo."
          readOnly={false}
          rows={null}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="sc-hzhJZQ gQJJDh"
    />
    <div
      className="sc-cWSHoV iHJLlN"
      id="campoPlacas"
    >
      <div
        className="sc-kpDqfm sc-eBMEME bTCkFB hwrltx"
        id="campoPlacas-etiqueta"
      >
        Placas
      </div>
      <div
        className="sc-kpDqfm sc-bmzYkS bTCkFB hLmFXM"
        id="campoPlacas-contenedorCampoDeTexto"
      >
        <input
          autoFocus={false}
          className="sc-iHGNWf jOstXH"
          data-indice={null}
          disabled={false}
          id="campoPlacas-campoDeTexto"
          maxLength={10}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyPress={[Function]}
          placeholder="Ingresa el número de placas."
          readOnly={false}
          rows={null}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="sc-ikkxIA eLqOjr"
    >
      <div>
        Soy un contenedor simulado
      </div>
    </div>
  </div>
</div>
`;
